src/Nethermind/Nethermind.Era1/EraWriter.cs

```diff
@@ -0,0 +1,225 @@
+using System.Text;
+using System.Threading.Tasks;
+using Cortex.SimpleSerialize;
+using Nethermind.Core;
+using Nethermind.Core.Crypto;
+using Nethermind.Core.Specs;
+using Nethermind.Int256;
+using Nethermind.Serialization.Rlp;
+using Nethermind.Serialization.Ssz;
+using Snappier;
+
+namespace Nethermind.Era1;
+internal class EraWriter : IDisposable
+{
+    public const int MaxEra1Size = 8192;
+
+    private long _startNumber;
+    private bool _firstBlock = true;
+    private long _totalWritten;
+    private readonly List<long> _entryIndexes = new();
+
+    private readonly HeaderDecoder _headerDecoder = new();
+    private readonly BlockBodyDecoder _blockBodyDecoder = new();
+    private readonly ReceiptMessageDecoder _receiptDecoder = new();
+
+    private readonly E2Store _e2Store;
+    private readonly AccumulatorCalculator _accumulatorCalculator;
+    private readonly ISpecProvider _specProvider;
+    private bool _disposedValue;
+    private bool _finalized;
+
+    public static EraWriter Create(string path, ISpecProvider specProvider)
+    {
+        if (string.IsNullOrWhiteSpace(path)) throw new ArgumentException($"'{nameof(path)}' cannot be null or whitespace.", nameof(path));
+        return Create(new FileStream(path, FileMode.Create), specProvider);
+    }
+    public static EraWriter Create(Stream stream, ISpecProvider specProvider)
+    {
+        if (specProvider is null) throw new ArgumentNullException(nameof(specProvider));
+
+        EraWriter b = new(new E2Store(stream), specProvider);
+        return b;
+    }
+
+    private EraWriter(E2Store e2Store, ISpecProvider specProvider)
+    {
+        _e2Store = e2Store;
+        _accumulatorCalculator = new();
+        _specProvider = specProvider;
+    }
+
+    public Task<bool> Add(Block block, TxReceipt[] receipts, CancellationToken cancellation = default)
+    {
+        if (block.TotalDifficulty == null)
+            throw new ArgumentException($"The block must have a {nameof(block.TotalDifficulty)}.", nameof(block));
+        return Add(block, receipts, block.TotalDifficulty.Value, cancellation);
+    }
+    public Task<bool> Add(Block block, TxReceipt[] receipts, UInt256 totalDifficulty, CancellationToken cancellation = default)
+    {
+        if (block.Header == null)
+            throw new ArgumentException("The block must have a header.", nameof(block));
+        if (block.Hash == null)
+            throw new ArgumentException("The block must have a hash.", nameof(block));
+
+        Rlp encodedHeader = _headerDecoder.Encode(block.Header);
+        Rlp encodedBody = _blockBodyDecoder.Encode(block.Body);
+        //Geth implementation has a byte array representing both TxPostState and StatusCode, and serializes whatever is set
+        RlpBehaviors behaviors = _specProvider.GetSpec(block.Header).IsEip658Enabled ? RlpBehaviors.Eip658Receipts : RlpBehaviors.None;
+        Rlp encodedReceipts = _receiptDecoder.Encode(receipts, behaviors);
+
```

Comment:
Use NettyRlpStream to not allocate memory for buffers.

Reply:
Changed to NettyRlpStream